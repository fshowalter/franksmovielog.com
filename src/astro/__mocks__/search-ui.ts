import type { PagefindAPI, PagefindResult } from "~/astro/search-ui";

import { debounce } from "~/utils/debounce";

// Mock SearchUI that works with our mocked pagefind API
export class SearchUI {
  private abortController: AbortController | undefined = undefined;
  private allResults: PagefindResult[] = [];
  private api: PagefindAPI | undefined = undefined;
  private clearButton: HTMLButtonElement | undefined = undefined;
  private container: HTMLElement | undefined = undefined;
  private readonly debouncedSearch: (query: string) => void;
  private displayedResults = 0;
  private initialized = false;
  private input: HTMLInputElement | undefined = undefined;
  private loadMoreButton: HTMLButtonElement | undefined = undefined;
  private loadMoreWrapper: HTMLElement | undefined = undefined;
  private resultsContainer: HTMLElement | undefined = undefined;
  private resultsCounter: HTMLElement | undefined = undefined;

  constructor() {
    // Create debounced search function
    this.debouncedSearch = debounce((query: string) => {
      void this.handleSearch(query);
    }, 150);
  }

  async destroy(): Promise<void> {
    if (this.abortController) {
      this.abortController.abort();
    }
    if (this.api && typeof this.api === "object" && "destroy" in this.api) {
      await this.api.destroy();
    }
    this.initialized = false;
  }

  async init(): Promise<void> {
    if (this.initialized) return;

    // Get DOM elements - look for the actual IDs in AstroPageShell
    this.input = document.querySelector("#pagefind-search-input");
    this.clearButton = document.querySelector("#pagefind-clear-button");
    this.resultsContainer = document.querySelector("#pagefind-results");
    this.resultsCounter = document.querySelector("#pagefind-results-counter");
    this.loadMoreWrapper = document.querySelector("#pagefind-load-more-wrapper");
    this.loadMoreButton = document.querySelector("#pagefind-load-more");

    // Import mocked pagefind module
    const pagefindModule = (await import(
      "/pagefind/pagefind.js"
    )) as PagefindAPI;
    this.api = pagefindModule;

    if (this.api && typeof this.api === "object" && "init" in this.api) {
      await this.api.init();
    }

    // Set up event listeners
    if (this.input) {
      this.input.addEventListener("input", (e) => {
        const target = e.target as HTMLInputElement;
        this.debouncedSearch(target.value);
      });
    }

    if (this.clearButton) {
      this.clearButton.addEventListener("click", () => {
        this.clearSearch();
      });
    }

    if (this.loadMoreButton) {
      this.loadMoreButton.addEventListener("click", () => {
        void this.loadMoreResults();
      });
    }

    this.initialized = true;
  }

  private clearSearch(): void {
    if (this.input) {
      this.input.value = "";
    }
    if (this.clearButton) {
      this.clearButton.classList.add("hidden");
    }
    if (this.resultsContainer) {
      this.resultsContainer.innerHTML = "";
    }
    if (this.resultsCounter) {
      this.resultsCounter.textContent = "";
    }
    if (this.loadMoreWrapper) {
      this.loadMoreWrapper.classList.add("hidden");
    }
    this.allResults = [];
    this.displayedResults = 0;

    // Focus input when clearing
    if (this.input) {
      this.input.focus();
    }
  }

  private getLoadingSkeleton(): string {
    return `
      <div class="animate-pulse">
        <div class="bg-subtle h-20 mb-4 rounded"></div>
        <div class="bg-subtle h-20 mb-4 rounded"></div>
        <div class="bg-subtle h-20 mb-4 rounded"></div>
      </div>
    `;
  }

  private async handleSearch(query: string): Promise<void> {
    // Cancel any pending search
    if (this.abortController) {
      this.abortController.abort();
    }

    const trimmedQuery = query.trim();

    // Update clear button visibility
    if (this.clearButton) {
      this.clearButton.classList.toggle("hidden", !trimmedQuery);
    }

    if (!trimmedQuery) {
      this.clearSearch();
      return;
    }

    // Create new abort controller for this search
    this.abortController = new AbortController();

    // Show loading skeleton
    if (this.resultsContainer) {
      this.resultsContainer.innerHTML = this.getLoadingSkeleton();
    }

    try {
      // Use the mocked debouncedSearch
      const searchResults = await this.api!.debouncedSearch(trimmedQuery);

      if (this.abortController.signal.aborted) {
        return;
      }

      this.allResults = searchResults.results || [];
      this.displayedResults = 0;

      // Update results counter
      if (this.resultsCounter) {
        const count = searchResults.unfilteredResultCount || 0;
        if (count === 0) {
          this.resultsCounter.textContent = `No results for "${trimmedQuery}"`;
        } else if (count === 1) {
          this.resultsCounter.textContent = `1 result for "${trimmedQuery}"`;
        } else {
          this.resultsCounter.textContent = `${count} results for "${trimmedQuery}"`;
        }
      }

      // Render initial results
      await this.renderResults();
    } catch (error: unknown) {
      if (error instanceof Error && error.name === "AbortError") {
        return; // Search was cancelled, ignore
      }
      console.error("Search failed:", error);
      if (this.resultsContainer) {
        this.resultsContainer.innerHTML = `
          <div class="text-muted">
            Search failed. Please try again.
          </div>
        `;
      }
    }
  }

  private async loadMoreResults(): Promise<void> {
    await this.renderResults();
  }

  private async renderResults(): Promise<void> {
    if (!this.resultsContainer) return;

    const resultsToShow = 5;
    const nextBatch = this.allResults.slice(
      this.displayedResults,
      this.displayedResults + resultsToShow,
    );

    if (this.displayedResults === 0) {
      // First render, replace skeleton
      this.resultsContainer.innerHTML = "";
    }

    if (nextBatch.length === 0 && this.displayedResults === 0) {
      this.resultsContainer.innerHTML = `
        <div class="text-muted">
          No results found. Try adjusting your search terms.
        </div>
      `;
      if (this.loadMoreWrapper) {
        this.loadMoreWrapper.classList.add("hidden");
      }
      return;
    }

    // Render results
    for (const result of nextBatch) {
      const data = await result.data();
      const resultHtml = `
        <div class="search-result">
          ${data.meta.image ? `<img src="${data.meta.image}" alt="${data.meta.image_alt || ""}" />` : ""}
          <div>
            <a href="${data.url}">${data.meta.title}</a>
            <p>${data.excerpt}</p>
          </div>
        </div>
      `;
      this.resultsContainer.insertAdjacentHTML("beforeend", resultHtml);
    }

    this.displayedResults += nextBatch.length;

    // Update load more button
    if (this.loadMoreWrapper && this.loadMoreButton) {
      const remaining = this.allResults.length - this.displayedResults;
      if (remaining > 0) {
        const nextCount = Math.min(remaining, resultsToShow);
        this.loadMoreButton.textContent = `Load ${nextCount} more (${remaining} total)`;
        this.loadMoreWrapper.classList.remove("hidden");
      } else {
        this.loadMoreWrapper.classList.add("hidden");
      }
    }
  }
}
