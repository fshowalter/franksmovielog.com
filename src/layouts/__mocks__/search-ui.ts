import { vi } from "vitest";

// Mock implementation that stubs out SearchAPI but keeps SearchUI functionality
// eslint-disable-next-line no-restricted-imports -- Mock needs to import the actual module
import type * as SearchUIModule from "../search-ui";

// Import the actual module to get the real SearchUI
const actual = await vi.importActual<typeof SearchUIModule>("../search-ui");

// Create a mock SearchAPI that doesn't require Pagefind
class MockSearchAPI {
  async destroy(): Promise<void> {
    // No-op
  }

  async init(): Promise<void> {
    // No-op - don't try to load Pagefind
  }

  search(): Promise<{ results: unknown[] }> {
    // Return empty results for testing
    return Promise.resolve({ results: [] });
  }
}

// Export the real SearchUI but with our mocked SearchAPI
export class SearchUI extends actual.SearchUI {
  constructor() {
    super();
    // Replace the API with our mock
    // @ts-expect-error - accessing private property for testing
    this.api = new MockSearchAPI();
  }
}
