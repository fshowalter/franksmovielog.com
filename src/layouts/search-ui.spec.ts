/**
 * @vitest-environment jsdom
 */
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

import { SearchUI } from "./search-ui";

// Mock the debounce utility
vi.mock("~/utils/debounce", () => ({
  debounce: <T extends (...args: unknown[]) => unknown>(fn: T) => fn,
}));

function createMockResult(id: string) {
  return {
    data: vi.fn().mockResolvedValue({
      excerpt: `Excerpt ${id}`,
      filters: {},
      meta: { title: `Result ${id}` },
      url: `/result${id}`,
      weighted_locations: [],
    }),
    id,
    score: 1,
    words: [1],
  };
}

// Helper functions for creating mock results
function createMockResultWithTitle(id: string, title: string) {
  return {
    data: vi.fn().mockResolvedValue({
      excerpt: `Excerpt ${id}`,
      filters: {},
      meta: { title },
      url: `/result${id}`,
      weighted_locations: [],
    }),
    id,
    score: 1,
    words: [1],
  };
}

describe("SearchUI", () => {
  let searchUI: SearchUI;
  let container: HTMLElement;
  let input: HTMLInputElement;
  let clearButton: HTMLButtonElement;
  let resultsContainer: HTMLElement;
  let resultsCounter: HTMLElement;
  let consoleErrorSpy: ReturnType<typeof vi.spyOn>;

  // Mock pagefind API
  const mockPagefindAPI = {
    debouncedSearch: vi.fn(),
    destroy: vi.fn(() => Promise.resolve()),
    filters: vi.fn().mockResolvedValue({}),
    init: vi.fn(() => Promise.resolve()),
    options: vi.fn(() => Promise.resolve()),
    preload: vi.fn(() => Promise.resolve()),
    search: vi.fn(),
  };

  beforeEach(() => {
    // Set up DOM structure
    document.body.innerHTML = `
      <div id="pagefind__search">
        <input class="pagefind-ui__search-input" type="text" />
        <button class="pagefind-ui__search-clear hidden">Clear</button>
        <div class="pagefind-ui__results-count"></div>
        <div class="pagefind-ui__results"></div>
        <div class="pagefind-ui__results-footer hidden">
          <button class="pagefind-ui__button">Load More</button>
        </div>
        <div class="pagefind-ui__filters"></div>
      </div>
    `;

    container = document.querySelector("#pagefind__search") as HTMLElement;
    input = container.querySelector(
      ".pagefind-ui__search-input",
    ) as HTMLInputElement;
    clearButton = container.querySelector(
      ".pagefind-ui__search-clear",
    ) as HTMLButtonElement;
    resultsContainer = container.querySelector(
      ".pagefind-ui__results",
    ) as HTMLElement;
    resultsCounter = container.querySelector(
      ".pagefind-ui__results-count",
    ) as HTMLElement;

    // Reset mocks
    vi.clearAllMocks();

    // Mock console.error to suppress expected error logs
    consoleErrorSpy = vi.spyOn(console, "error").mockImplementation(() => {
      // Explicitly do nothing - suppress output
    });

    // Set up default mock return values
    mockPagefindAPI.search.mockResolvedValue({
      filters: {},
      results: [],
      timings: { preload: 0, search: 0, total: 0 },
      totalFilters: {},
      unfilteredResultCount: 0,
    });

    mockPagefindAPI.debouncedSearch.mockResolvedValue({
      filters: {},
      results: [],
      timings: { preload: 0, search: 0, total: 0 },
      totalFilters: {},
      unfilteredResultCount: 0,
    });

    // Mock dynamic import for pagefind
    vi.doMock("/pagefind/pagefind.js", () => mockPagefindAPI);

    searchUI = new SearchUI();
  });

  afterEach(() => {
    document.body.innerHTML = "";
    consoleErrorSpy.mockRestore();
    vi.clearAllMocks();
    vi.doUnmock("/pagefind/pagefind.js");
  });

  describe("initialization", () => {
    it("should handle missing DOM elements gracefully", async () => {
      document.body.innerHTML = "";
      const newSearchUI = new SearchUI();

      // Should not throw when elements are missing
      await expect(newSearchUI.init()).resolves.toBeUndefined();
    });

    it("should initialize with fallback element selectors", async () => {
      // Set up alternative DOM structure (without container)
      document.body.innerHTML = `
        <input id="pagefind-search-input" type="text" />
        <button id="pagefind-clear-button" class="hidden">Clear</button>
        <div id="pagefind-results-counter"></div>
        <div id="pagefind-results"></div>
        <div id="pagefind-load-more-wrapper" class="hidden">
          <button id="pagefind-load-more">Load More</button>
        </div>
      `;

      const newSearchUI = new SearchUI();
      await newSearchUI.init();

      const fallbackInput = document.querySelector(
        "#pagefind-search-input",
      ) as HTMLInputElement;
      expect(fallbackInput).toBeTruthy();
    });
  });

  describe("search functionality", () => {
    it("should clear search when clear button is clicked", async () => {
      await searchUI.init();

      // First set some input
      input.value = "test";
      input.dispatchEvent(new Event("input"));

      expect(clearButton.classList.contains("hidden")).toBe(false);

      // Click clear button
      clearButton.click();

      expect(input.value).toBe("");
      expect(clearButton.classList.contains("hidden")).toBe(true);
      expect(resultsContainer.innerHTML).toBe("");
      expect(resultsCounter.textContent).toBe("");
    });

    it("should clear search when input is emptied", async () => {
      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      input.value = "";
      input.dispatchEvent(new Event("input"));

      expect(clearButton.classList.contains("hidden")).toBe(true);
      expect(resultsContainer.innerHTML).toBe("");
    });

    it("should cancel pending search when new search starts", async () => {
      await searchUI.init();

      const abortSpy = vi.spyOn(AbortController.prototype, "abort");

      // Start first search
      input.value = "first";
      input.dispatchEvent(new Event("input"));

      // Start second search before first completes
      input.value = "second";
      input.dispatchEvent(new Event("input"));

      expect(abortSpy).toHaveBeenCalled();
      abortSpy.mockRestore();
    });
  });

  describe("destroy functionality", () => {
    it("should abort pending searches when destroyed", async () => {
      // Initialize first to set up abort controller
      await searchUI.init();

      const abortSpy = vi.spyOn(AbortController.prototype, "abort");

      // Start a search
      input.value = "test";
      input.dispatchEvent(new Event("input"));

      // Wait a tick for the search to start
      await new Promise((resolve) => setTimeout(resolve, 0));

      // Destroy before search completes
      await searchUI.destroy();

      expect(abortSpy).toHaveBeenCalled();
      abortSpy.mockRestore();
    });
  });

  describe("rendering", () => {
    it("should trim search query", async () => {
      await searchUI.init();

      // Input with whitespace
      input.value = "  ";
      input.dispatchEvent(new Event("input"));

      // Should clear for whitespace-only input
      await new Promise((resolve) => setTimeout(resolve, 10));
      expect(resultsContainer.innerHTML).toBe("");
    });

    it("should show/hide clear button based on input", async () => {
      await searchUI.init();

      // Initially hidden
      expect(clearButton.classList.contains("hidden")).toBe(true);

      // Show when input has value
      input.value = "test";
      input.dispatchEvent(new Event("input"));
      expect(clearButton.classList.contains("hidden")).toBe(false);

      // Hide when input is cleared
      input.value = "";
      input.dispatchEvent(new Event("input"));
      expect(clearButton.classList.contains("hidden")).toBe(true);
    });
  });

  describe("search results", () => {
    it("should handle search results with data", async () => {
      const mockResults = [
        {
          data: vi.fn().mockResolvedValue({
            excerpt: "Test excerpt 1",
            filters: {},
            meta: { title: "Test Title 1" },
            url: "/test1",
            weighted_locations: [],
          }),
          id: "1",
          score: 1,
          words: [1],
        },
        {
          data: vi.fn().mockResolvedValue({
            excerpt: "Test excerpt 2",
            filters: {},
            meta: { title: "Test Title 2" },
            url: "/test2",
            weighted_locations: [],
          }),
          id: "2",
          score: 0.9,
          words: [1],
        },
      ];

      mockPagefindAPI.debouncedSearch.mockResolvedValueOnce({
        filters: {},
        results: mockResults,
        timings: { preload: 0, search: 0, total: 0 },
        totalFilters: {},
        unfilteredResultCount: 2,
      });

      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(resultsContainer.innerHTML).toContain("Test Title 1");
      expect(resultsContainer.innerHTML).toContain("Test excerpt 1");
      expect(resultsContainer.innerHTML).toContain("/test1");
      expect(resultsCounter.textContent).toContain("2 results");
      expect(resultsCounter.textContent).toContain("test");
    });

    it("should render results with images when available", async () => {
      const mockResults = [
        {
          data: vi.fn().mockResolvedValue({
            excerpt: "Test excerpt",
            filters: {},
            meta: {
              image: "/test-image.jpg",
              image_alt: "Test alt text",
              title: "Test Title",
            },
            url: "/test",
            weighted_locations: [],
          }),
          id: "1",
          score: 1,
          words: [1],
        },
      ];

      mockPagefindAPI.debouncedSearch.mockResolvedValueOnce({
        filters: {},
        results: mockResults,
        timings: { preload: 0, search: 0, total: 0 },
        totalFilters: {},
        unfilteredResultCount: 1,
      });

      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(resultsContainer.innerHTML).toContain('src="/test-image.jpg"');
      expect(resultsContainer.innerHTML).toContain('alt="Test alt text"');
    });

    it("should show correct message for zero results", async () => {
      mockPagefindAPI.debouncedSearch.mockResolvedValueOnce({
        filters: {},
        results: [],
        timings: { preload: 0, search: 0, total: 0 },
        totalFilters: {},
        unfilteredResultCount: 0,
      });

      await searchUI.init();

      input.value = "notfound";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(resultsCounter.textContent).toBe('No results for "notfound"');
      expect(resultsContainer.innerHTML).toContain(
        "No results found. Try adjusting your search terms.",
      );
    });

    it("should show correct message for single result", async () => {
      const mockResults = [
        {
          data: vi.fn().mockResolvedValue({
            excerpt: "Test",
            filters: {},
            meta: { title: "Single Result" },
            url: "/single",
            weighted_locations: [],
          }),
          id: "1",
          score: 1,
          words: [1],
        },
      ];

      mockPagefindAPI.debouncedSearch.mockResolvedValueOnce({
        filters: {},
        results: mockResults,
        timings: { preload: 0, search: 0, total: 0 },
        totalFilters: {},
        unfilteredResultCount: 1,
      });

      await searchUI.init();

      input.value = "single";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(resultsCounter.textContent).toBe('1 result for "single"');
    });
  });

  describe("load more functionality", () => {
    it("should load more results when button is clicked", async () => {
      const mockResults = Array.from({ length: 10 }, (_, i) =>
        createMockResultWithTitle(`${i + 1}`, `Result ${i + 1}`),
      );

      mockPagefindAPI.debouncedSearch.mockResolvedValueOnce({
        filters: {},
        results: mockResults,
        timings: { preload: 0, search: 0, total: 0 },
        totalFilters: {},
        unfilteredResultCount: 10,
      });

      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      const loadMoreWrapper = container.querySelector(
        ".pagefind-ui__results-footer",
      ) as HTMLElement;
      const loadMoreButton = container.querySelector(
        ".pagefind-ui__button",
      ) as HTMLButtonElement;

      expect(loadMoreWrapper.classList.contains("hidden")).toBe(false);
      expect(loadMoreButton.textContent).toContain("Load 5 more");
      expect(loadMoreButton.textContent).toContain("5 total");

      expect(resultsContainer.innerHTML).toContain("Result 1");
      expect(resultsContainer.innerHTML).toContain("Result 5");
      expect(resultsContainer.innerHTML).not.toContain("Result 6");

      loadMoreButton.click();
      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(resultsContainer.innerHTML).toContain("Result 6");
      expect(resultsContainer.innerHTML).toContain("Result 10");
      expect(loadMoreWrapper.classList.contains("hidden")).toBe(true);
    });

    it("should handle load more errors gracefully", async () => {
      const createMockResult = (id: string, shouldFail = false) => ({
        data: shouldFail
          ? vi.fn().mockRejectedValue(new Error("Failed to load"))
          : vi.fn().mockResolvedValue({
              excerpt: `Excerpt ${id}`,
              filters: {},
              meta: { title: `Result ${id}` },
              url: `/result${id}`,
              weighted_locations: [],
            }),
        id,
        score: 1,
        words: [1],
      });

      const mockResults = [
        ...Array.from({ length: 5 }, (_, i) =>
          createMockResult(`${i + 1}`, false),
        ),
        ...Array.from({ length: 5 }, (_, i) =>
          createMockResult(`${i + 6}`, true),
        ),
      ];

      mockPagefindAPI.debouncedSearch.mockResolvedValueOnce({
        filters: {},
        results: mockResults,
        timings: { preload: 0, search: 0, total: 0 },
        totalFilters: {},
        unfilteredResultCount: 10,
      });

      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      const loadMoreButton = container.querySelector(
        ".pagefind-ui__button",
      ) as HTMLButtonElement;

      // Clear any previous console.error calls
      consoleErrorSpy.mockClear();

      loadMoreButton.click();
      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(consoleErrorSpy).toHaveBeenCalledWith(
        "Failed to load more results:",
        expect.any(Error),
      );
      expect(resultsContainer.innerHTML).toContain(
        "Failed to load more results.",
      );
    });

    it("should create screen reader announcements when loading more", async () => {
      const mockResults = Array.from({ length: 10 }, (_, i) =>
        createMockResult(`${i + 1}`),
      );

      mockPagefindAPI.debouncedSearch.mockResolvedValueOnce({
        filters: {},
        results: mockResults,
        timings: { preload: 0, search: 0, total: 0 },
        totalFilters: {},
        unfilteredResultCount: 10,
      });

      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      const loadMoreButton = container.querySelector(
        ".pagefind-ui__button",
      ) as HTMLButtonElement;

      loadMoreButton.click();
      await new Promise((resolve) => setTimeout(resolve, 50));

      const announcement = document.querySelector('[role="status"]');
      expect(announcement).toBeTruthy();
      expect(announcement?.getAttribute("aria-live")).toBe("polite");
      expect(announcement?.classList.contains("sr-only")).toBe(true);
      expect(announcement?.textContent).toContain("5 more results loaded");

      await new Promise((resolve) => setTimeout(resolve, 1100));
      expect(document.querySelector('[role="status"]')).toBeFalsy();
    });
  });

  describe("error handling", () => {
    it("should handle initialization errors", async () => {
      // Override the mock to simulate a broken pagefind module
      vi.doUnmock("/pagefind/pagefind.js");
      vi.doMock("/pagefind/pagefind.js", () => ({
        // Return an object without the required 'options' method
        // This will cause an error when trying to call pagefindModule.options
        init: vi.fn(),
        search: vi.fn(),
      }));

      const errorSearchUI = new SearchUI();
      await errorSearchUI.init();

      expect(consoleErrorSpy).toHaveBeenCalledWith(
        "Failed to initialize search API:",
        expect.any(Error),
      );

      // Restore the mock for other tests
      vi.doUnmock("/pagefind/pagefind.js");
      vi.doMock("/pagefind/pagefind.js", () => mockPagefindAPI);
    });

    it("should handle search errors gracefully", async () => {
      mockPagefindAPI.debouncedSearch.mockRejectedValueOnce(
        new Error("Network error"),
      );

      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(consoleErrorSpy).toHaveBeenCalledWith(
        "Search failed:",
        expect.any(Error),
      );
      expect(resultsContainer.innerHTML).toContain(
        "Search failed. Please try again.",
      );
    });

    it("should ignore aborted search errors", async () => {
      await searchUI.init();

      // Clear any previous calls to console.error
      consoleErrorSpy.mockClear();

      // Set up the mock to reject with an AbortError
      const abortError = new Error("Aborted");
      abortError.name = "AbortError";
      mockPagefindAPI.debouncedSearch.mockRejectedValueOnce(abortError);

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(consoleErrorSpy).not.toHaveBeenCalled();
      expect(resultsContainer.innerHTML).not.toContain("Search failed");
    });

    it("should show loading skeleton while searching", async () => {
      let resolveSearch: (value: unknown) => void;
      const searchPromise = new Promise((resolve) => {
        resolveSearch = resolve;
      });

      mockPagefindAPI.debouncedSearch.mockReturnValueOnce(searchPromise);

      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 10));

      expect(resultsContainer.innerHTML).toContain("animate-pulse");
      expect(resultsContainer.innerHTML).toContain("bg-subtle");

      resolveSearch!({
        filters: {},
        results: [],
        timings: { preload: 0, search: 0, total: 0 },
        totalFilters: {},
        unfilteredResultCount: 0,
      });

      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(resultsContainer.innerHTML).not.toContain("animate-pulse");
    });
  });

  describe("edge cases", () => {
    it("should not reinitialize when already initialized", async () => {
      await searchUI.init();

      // Clear the mock call count after first init
      mockPagefindAPI.init.mockClear();

      await searchUI.init();

      // Should not call init again
      expect(mockPagefindAPI.init).not.toHaveBeenCalled();
    });

    it("should focus input when clear button is clicked", async () => {
      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      const focusSpy = vi.spyOn(input, "focus");

      clearButton.click();

      expect(focusSpy).toHaveBeenCalled();
      focusSpy.mockRestore();
    });

    it("should maintain scroll position when loading more results", async () => {
      const mockResults = Array.from({ length: 10 }, (_, i) =>
        createMockResult(`${i + 1}`),
      );

      mockPagefindAPI.debouncedSearch.mockResolvedValueOnce({
        filters: {},
        results: mockResults,
        timings: { preload: 0, search: 0, total: 0 },
        totalFilters: {},
        unfilteredResultCount: 10,
      });

      await searchUI.init();

      input.value = "test";
      input.dispatchEvent(new Event("input"));

      await new Promise((resolve) => setTimeout(resolve, 50));

      resultsContainer.scrollTop = 100;

      const loadMoreButton = container.querySelector(
        ".pagefind-ui__button",
      ) as HTMLButtonElement;

      loadMoreButton.click();
      await new Promise((resolve) => setTimeout(resolve, 50));

      expect(resultsContainer.scrollTop).toBe(100);
    });

    it("should properly clean up on destroy", async () => {
      await searchUI.init();

      // Start a search but don't wait for it to complete
      input.value = "test";
      input.dispatchEvent(new Event("input"));

      // Immediately destroy while search is pending
      await searchUI.destroy();

      expect(mockPagefindAPI.destroy).toHaveBeenCalled();
      // The destroy method clears internal state but doesn't update DOM
      // This is acceptable as destroy is meant to clean up resources
    });
  });
});
