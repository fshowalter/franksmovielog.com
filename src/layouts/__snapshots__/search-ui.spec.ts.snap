// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchUI > result rendering > should render result item with image 1`] = `
"
        <li class="gap-x-6 tablet:px-6 laptop:px-8 py-6 px-[8%] hover:bg-subtle border-t border-default last-of-type:border-b grid grid-cols-[min(25%,80px)_1fr] focus-within:bg-subtle focus-within:outline-[rgb(38,132,255)] focus-within:outline-1 focus-within:-outline-offset-2">
          
            <div class="shrink-0 drop-shadow-md">
              <img 
                src="/test.jpg" 
                alt="Test Alt"
                class="h-auto w-full"
                loading="lazy"
              >
            </div>
          
          <div class="min-w-0 flex-1">
            <h3 class="font-sans text-base font-semibold text-accent">
              <a href="/test" class="block">
                Test Title
              </a>
            </h3>
            <p class="text-sm leading-normal text-default">
              Test excerpt
            </p>
          </div>
        </li>
    "
`;
